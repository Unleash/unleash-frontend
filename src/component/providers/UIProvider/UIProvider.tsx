import React, { useState } from 'react';
import UIContext, { createEmptyToast } from '../../../contexts/UIContext';
import { IToast } from '../../../interfaces/toast';

const UIProvider: React.FC = ({ children }) => {
    const [toastData, setToast] = useState<IToast>(createEmptyToast());
    const [showFeedback, setShowFeedback] = useState(false);

    const context = React.useMemo(
        () => ({
            setToast,
            toastData,
            showFeedback,
            setShowFeedback,
        }),
        [toastData, showFeedback]
    );

    return <UIContext.Provider value={context}>{children}</UIContext.Provider>;
};

export default UIProvider;
