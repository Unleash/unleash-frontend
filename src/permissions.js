export const ADMIN = 'ADMIN';
export const CREATE_FEATURE = 'CREATE_FEATURE';
export const UPDATE_FEATURE = 'UPDATE_FEATURE';
export const DELETE_FEATURE = 'DELETE_FEATURE';
export const CREATE_STRATEGY = 'CREATE_STRATEGY';
export const UPDATE_STRATEGY = 'UPDATE_STRATEGY';
export const DELETE_STRATEGY = 'DELETE_STRATEGY';
export const UPDATE_APPLICATION = 'UPDATE_APPLICATION';
export const CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
export const UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
export const DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';

export function hasPermission(user, permission) {
    return (
        user &&
        (!user.permissions || user.permissions.indexOf(ADMIN) !== -1 || user.permissions.indexOf(permission) !== -1)
    );
}
