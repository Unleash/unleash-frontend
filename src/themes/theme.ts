import { createTheme } from '@mui/material/styles';
import { colors } from './colors';

export default createTheme({
    boxShadows: {
        main: '0px 2px 4px rgba(129, 122, 254, 0.2)',
    },
    typography: {
        fontFamily: 'Sen, Roboto, sans-serif',
        fontWeightBold: '700',
        fontWeightMedium: '700',
        allVariants: { lineHeight: 1.4 },
        button: { lineHeight: 1.75 },
        h1: {
            fontSize: '1.5rem',
            lineHeight: 1.875,
        },
    },
    fontSizes: {
        mainHeader: '1.25rem',
        subHeader: '1.1rem',
        bodySize: '1rem',
        smallBody: '0.9rem',
        smallerBody: '0.8rem',
    },
    fontWeight: {
        thin: 300,
        medium: 400,
        semi: 700,
        bold: 700,
    },
    shape: {
        borderRadius: '3px',
        borderRadiusLarge: '10px',
        borderRadiusExtraLarge: '25px',
    },
    palette: {
        primary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        secondary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        info: {
            light: colors.blue[700],
            main: colors.blue[700],
            dark: colors.blue[800],
        },
        success: {
            light: colors.green[700],
            main: colors.green[700],
            dark: colors.green[800],
        },
        warning: {
            light: colors.orange[700],
            main: colors.orange[700],
            dark: colors.orange[800],
        },
        error: {
            light: colors.red[700],
            main: colors.red[700],
            dark: colors.red[800],
        },
        divider: colors.grey[300],
        dividerAlternative: colors.grey[500],
        grey: colors.grey,
        text: {
            primary: colors.grey[900],
            secondary: colors.grey[800],
            disabled: colors.grey[600],
        },
        code: {
            main: '#0b8c8f',
            diffAdd: 'green',
            diffSub: 'red',
            diffNeutral: 'black',
            edited: 'blue',
            background: '#efefef',
        },
        activityIndicators: {
            unknown: colors.grey[100],
            recent: colors.green[100],
            inactive: colors.orange[200],
            abandoned: colors.red[200],
        },
    },
    components: {
        MuiTableHead: {
            styleOverrides: {
                root: {
                    background: colors.grey[200],
                },
            },
        },
        MuiLink: {
            styleOverrides: {
                root: {
                    color: colors.purple[900],
                },
            },
        },
        MuiBreadcrumbs: {
            styleOverrides: {
                root: {
                    '& a': {
                        color: colors.purple[900],
                    },
                },
            },
        },
        MuiTableRow: {
            styleOverrides: {
                root: {
                    '&.MuiTableRow-hover:hover': {
                        background: colors.grey[100],
                    },
                },
            },
        },
    },
});
