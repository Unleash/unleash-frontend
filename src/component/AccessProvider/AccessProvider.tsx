import { FC } from "react";

import AccessContext from '../../contexts/AccessContext'
import { ADMIN } from "./permissions";

// TODO: Type up redux store
interface IAccessProvider {
    store: any;
}

interface IPermission {
  permission: string;
  project: string | null;
}

const AccessProvider: FC<IAccessProvider> = ({store, children}) => {
  const hasAccess = (permission: string, project: string) => {
    const permissions = store.getState().user.get('permissions') || [];

    const result = permissions.some((p: IPermission) => {
      if(p.permission === ADMIN) {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    });

    return result;
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export default AccessProvider;