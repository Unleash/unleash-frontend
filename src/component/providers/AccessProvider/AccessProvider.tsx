import { ReactElement, ReactNode, useCallback, useMemo } from 'react';
import AccessContext, { IAccessContext } from '../../../contexts/AccessContext';
import { ADMIN } from './permissions';
import { IPermission } from '../../../interfaces/user';
import { useAuthPermissions } from '../../../hooks/api/getters/useAuth/useAuthPermissions';

interface IAccessProviderProps {
    children: ReactNode;
    permissions: IPermission[];
}

// TODO(olav): Mock useAuth instead of using props.permissions in tests.
const AccessProvider = (props: IAccessProviderProps): ReactElement => {
    const auth = useAuthPermissions();
    const permissions = props.permissions ?? auth.permissions;

    const isAdmin: boolean = useMemo(() => {
        return checkAdmin(permissions);
    }, [permissions]);

    const hasAccess = useCallback(
        (permission: string, project?: string, environment?: string) => {
            return checkPermissions(
                permissions,
                permission,
                project,
                environment
            );
        },
        [permissions]
    );

    const value: IAccessContext = useMemo(
        () => ({
            isAdmin,
            hasAccess,
        }),
        [isAdmin, hasAccess]
    );

    return (
        <AccessContext.Provider value={value}>
            {props.children}
        </AccessContext.Provider>
    );
};

const checkAdmin = (permissions: IPermission[] | undefined): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return p.permission === ADMIN;
    });
};

const checkPermissions = (
    permissions: IPermission[] | undefined,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return checkPermission(p, permission, project, environment);
    });
};

const checkPermission = (
    p: IPermission,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permission) {
        console.warn(`Missing permission for AccessProvider: ${permission}`)
        return false
    }

    if (p.permission === ADMIN) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        (p.environment === environment || p.environment === '*')
    ) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        p.environment === null
    ) {
        return true;
    }

    return (
        p.permission === permission &&
        p.project === undefined &&
        p.environment === null
    );
};

export default AccessProvider;
