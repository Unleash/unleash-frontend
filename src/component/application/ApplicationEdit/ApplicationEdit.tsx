/* eslint react/no-multi-comp:off */
import React, { useContext, useState } from 'react';
import {
    Avatar,
    Icon,
    IconButton,
    LinearProgress,
    Link,
    Typography,
} from '@material-ui/core';
import { Link as LinkIcon } from '@material-ui/icons';
import ConditionallyRender from 'component/common/ConditionallyRender/ConditionallyRender';
import { UPDATE_APPLICATION } from 'component/providers/AccessProvider/permissions';
import { ApplicationView } from '../ApplicationView/ApplicationView';
import { ApplicationUpdate } from '../ApplicationUpdate/ApplicationUpdate';
import Dialogue from 'component/common/Dialogue';
import PageContent from 'component/common/PageContent';
import HeaderTitle from 'component/common/HeaderTitle';
import AccessContext from 'contexts/AccessContext';
import useApplicationsApi from 'hooks/api/actions/useApplicationsApi/useApplicationsApi';
import useApplication from 'hooks/api/getters/useApplication/useApplication';
import { useHistory, useParams } from 'react-router-dom';
import { useLocationSettings } from 'hooks/useLocationSettings';
import useToast from 'hooks/useToast';
import PermissionButton from 'component/common/PermissionButton/PermissionButton';
import { formatDateYMD } from 'utils/formatDate';
import { formatUnknownError } from 'utils/formatUnknownError';
import { TabNav } from 'component/common/TabNav/TabNav/TabNav';

export const ApplicationEdit = () => {
    const history = useHistory();
    const { name } = useParams<{ name: string }>();
    const { application, loading } = useApplication(name);
    const { appName, url, description, icon = 'apps', createdAt } = application;
    const { hasAccess } = useContext(AccessContext);
    const { deleteApplication } = useApplicationsApi();
    const { locationSettings } = useLocationSettings();
    const { setToastData, setToastApiError } = useToast();

    const [showDialog, setShowDialog] = useState(false);

    const toggleModal = () => {
        setShowDialog(!showDialog);
    };

    const formatDate = (v: string) => formatDateYMD(v, locationSettings.locale);

    const onDeleteApplication = async (evt: React.SyntheticEvent) => {
        evt.preventDefault();
        try {
            await deleteApplication(appName);
            setToastData({
                title: 'Deleted Successfully',
                text: 'Application deleted successfully',
                type: 'success',
            });
            history.push('/applications');
        } catch (error: unknown) {
            setToastApiError(formatUnknownError(error));
        }
    };

    const renderModal = () => (
        <Dialogue
            open={showDialog}
            onClose={toggleModal}
            onClick={onDeleteApplication}
            title="Are you sure you want to delete this application?"
        />
    );
    const tabData = [
        {
            label: 'Application overview',
            component: <ApplicationView />,
        },
        {
            label: 'Edit application',
            component: <ApplicationUpdate application={application} />,
        },
    ];

    if (loading) {
        return (
            <div>
                <p>Loading...</p>
                <LinearProgress />
            </div>
        );
    } else if (!application) {
        return <p>Application ({appName}) not found</p>;
    }
    return (
        <PageContent
            headerContent={
                <HeaderTitle
                    title={
                        <span
                            style={{
                                display: 'flex',
                                alignItems: 'center',
                            }}
                        >
                            <Avatar style={{ marginRight: '8px' }}>
                                <Icon>{icon || 'apps'}</Icon>
                            </Avatar>
                            {appName}
                        </span>
                    }
                    actions={
                        <>
                            <ConditionallyRender
                                condition={Boolean(url)}
                                show={
                                    <IconButton component={Link} href={url}>
                                        <LinkIcon titleAccess={url} />
                                    </IconButton>
                                }
                            />

                            <PermissionButton
                                tooltip="Delete application"
                                onClick={toggleModal}
                                permission={UPDATE_APPLICATION}
                            >
                                Delete
                            </PermissionButton>
                        </>
                    }
                />
            }
        >
            <div>
                <Typography variant="body1">{description || ''}</Typography>
                <Typography variant="body2">
                    Created: <strong>{formatDate(createdAt)}</strong>
                </Typography>
            </div>
            <ConditionallyRender
                condition={hasAccess(UPDATE_APPLICATION)}
                show={
                    <div>
                        {renderModal()}
                        <TabNav tabData={tabData} />
                    </div>
                }
            />
        </PageContent>
    );
};
