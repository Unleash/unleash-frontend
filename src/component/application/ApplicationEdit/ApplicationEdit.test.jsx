import { ThemeProvider } from '@material-ui/core';
import { ApplicationEdit } from './ApplicationEdit';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { ADMIN } from 'component/providers/AccessProvider/permissions';
import theme from 'themes/mainTheme';
import AccessProvider from 'component/providers/AccessProvider/AccessProvider';
import UIProvider from 'component/providers/UIProvider/UIProvider';

test('renders correctly if no application', () => {
    const tree = renderer
        .create(
            <AccessProvider permissions={[{ permission: ADMIN }]}>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <MemoryRouter initialEntries={['/test']}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                deleteApplication={jest.fn()}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </MemoryRouter>
                    </UIProvider>
                </ThemeProvider>
            </AccessProvider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('renders correctly without permission', () => {
    const tree = renderer
        .create(
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <AccessProvider permissions={[{ permission: ADMIN }]}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                deleteApplication={jest.fn()}
                                application={{
                                    appName: 'test-app',
                                    instances: [
                                        {
                                            instanceId: 'instance-1',
                                            clientIp: '123.123.123.123',
                                            lastSeen: '2017-02-23T15:56:49',
                                            sdkVersion: '4.0',
                                        },
                                    ],
                                    strategies: [
                                        {
                                            name: 'StrategyA',
                                            description: 'A description',
                                        },
                                        {
                                            name: 'StrategyB',
                                            description: 'B description',
                                            notFound: true,
                                        },
                                    ],
                                    seenToggles: [
                                        {
                                            name: 'ToggleA',
                                            description: 'this is A toggle',
                                            enabled: true,
                                            project: 'default',
                                        },
                                        {
                                            name: 'ToggleB',
                                            description: 'this is B toggle',
                                            enabled: false,
                                            notFound: true,
                                            project: 'default',
                                        },
                                    ],
                                    url: 'http://example.org',
                                    description: 'app description',
                                }}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </AccessProvider>
                    </UIProvider>
                </ThemeProvider>
            </MemoryRouter>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('renders correctly with permissions', () => {
    const tree = renderer
        .create(
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <AccessProvider permissions={[{ permission: ADMIN }]}>
                            <ApplicationEdit
                                fetchApplication={() => Promise.resolve({})}
                                storeApplicationMetaData={jest.fn()}
                                deleteApplication={jest.fn()}
                                application={{
                                    appName: 'test-app',
                                    instances: [
                                        {
                                            instanceId: 'instance-1',
                                            clientIp: '123.123.123.123',
                                            lastSeen: '2017-02-23T15:56:49',
                                            sdkVersion: '4.0',
                                        },
                                    ],
                                    strategies: [
                                        {
                                            name: 'StrategyA',
                                            description: 'A description',
                                        },
                                        {
                                            name: 'StrategyB',
                                            description: 'B description',
                                            notFound: true,
                                        },
                                    ],
                                    seenToggles: [
                                        {
                                            name: 'ToggleA',
                                            description: 'this is A toggle',
                                            enabled: true,
                                            project: 'default',
                                        },
                                        {
                                            name: 'ToggleB',
                                            description: 'this is B toggle',
                                            enabled: false,
                                            notFound: true,
                                            project: 'default',
                                        },
                                    ],
                                    url: 'http://example.org',
                                    description: 'app description',
                                }}
                                locationSettings={{ locale: 'en-GB' }}
                            />
                        </AccessProvider>
                    </UIProvider>
                </ThemeProvider>
            </MemoryRouter>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
