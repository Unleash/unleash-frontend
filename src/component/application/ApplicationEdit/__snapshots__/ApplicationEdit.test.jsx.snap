// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly if no application 1`] = `
<div>
  <p>
    Loading...
  </p>
  <span
    className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate mui-130qhmc-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate mui-yt7v2r-MuiLinearProgress-bar1"
      style={Object {}}
    />
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate mui-1wjhhwq-MuiLinearProgress-bar2"
      style={Object {}}
    />
  </span>
</div>
`;

exports[`renders correctly with permissions 1`] = `
<div>
  <p>
    Loading...
  </p>
  <span
    className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate mui-130qhmc-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate mui-yt7v2r-MuiLinearProgress-bar1"
      style={Object {}}
    />
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate mui-1wjhhwq-MuiLinearProgress-bar2"
      style={Object {}}
    />
  </span>
</div>
`;

exports[`renders correctly without permission 1`] = `
<div>
  <p>
    Loading...
  </p>
  <span
    className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate mui-130qhmc-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate mui-yt7v2r-MuiLinearProgress-bar1"
      style={Object {}}
    />
    <span
      className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate mui-1wjhhwq-MuiLinearProgress-bar2"
      style={Object {}}
    />
  </span>
</div>
`;
